diff --git a/akregator/src/actionmanagerimpl.cpp b/akregator/src/actionmanagerimpl.cpp
index 33a5a20..4a1306a 100644
--- a/akregator/src/actionmanagerimpl.cpp
+++ b/akregator/src/actionmanagerimpl.cpp
@@ -176,13 +176,10 @@ void ActionManagerImpl::initPart()
     action->setIcon(KIcon("document-export"));
     connect(action, SIGNAL(triggered(bool)), d->part, SLOT(fileExport()));
 
-    KStandardAction::configureNotifications(d->part, SLOT(showKNotifyOptions()), d->actionCollection); // options_configure_notifications
-
-    KStandardAction::keyBindings(d->part->parent(), SLOT(optionsConfigureKeys()), d->actionCollection); // options_configure_keybinding
-
-    KStandardAction::configureToolbars(d->part->parent(), SLOT(optionsConfigureToolbars()), d->actionCollection); // options_configure_toolbars
-
-    KStandardAction::preferences(d->part, SLOT(showOptions()), d->actionCollection); // options_configure
+    KAction *configure = d->actionCollection->addAction("options_configure");
+    configure->setText(i18n("&Configure Akregator..."));
+    configure->setIcon(KIcon("configure"));
+    connect(configure, SIGNAL(triggered()), d->part, SLOT(showOptions()));
 
     /*action = d->actionCollection->addAction("akregator_configure_akregator");
     action->setIcon(KIcon("configure"));
diff --git a/akregator/src/akregator_part.cpp b/akregator/src/akregator_part.cpp
index 2aefe1f..8fc0126 100644
--- a/akregator/src/akregator_part.cpp
+++ b/akregator/src/akregator_part.cpp
@@ -59,6 +59,7 @@
 #include <KParts/GenericFactory>
 #include <KParts/Plugin>
 #include <KCMultiDialog>
+#include <kstandardaction.h>
 
 #include <QFile>
 #include <QObject>
@@ -106,6 +107,7 @@ Part::Part( QWidget *parentWidget, QObject *parent, const QVariantList& )
     setPluginInterfaceVersion( AKREGATOR_PLUGIN_INTERFACE_VERSION );
 
     setComponentData( AkregatorFactory::componentData() );
+    setXMLFile("akregator_part.rc", true);
 
     new PartAdaptor( this );
     QDBusConnection::sessionBus().registerObject("/Akregator", this);
@@ -178,8 +180,6 @@ Part::Part( QWidget *parentWidget, QObject *parent, const QVariantList& )
     connect(m_autosaveTimer, SIGNAL(timeout()), this, SLOT(slotSaveFeedList()));
     m_autosaveTimer->start(5*60*1000); // 5 minutes
 
-    setXMLFile("akregator_part.rc", true);
-
     initFonts();
 
     QString useragent = QString( "Akregator/%1; syndication" ).arg( AKREGATOR_VERSION );
diff --git a/akregator/src/akregator_part.rc b/akregator/src/akregator_part.rc
index ccd8e20..787af4c 100644
--- a/akregator/src/akregator_part.rc
+++ b/akregator/src/akregator_part.rc
@@ -73,10 +73,7 @@
    <text>&amp;Settings</text>
    <Merge name="StandardToolBarMenuHandler" />
    <Action name="show_quick_filter" />
-   <Separator/>
-   <Action group="settings_configure" name="options_configure_keybinding" />
-   <Action group="settings_configure" name="options_configure_notifications" />
-   <Action group="settings_configure" name="options_configure_toolbars" />
+   <Merge/>
    <Action group="settings_configure" name="options_configure" />
   </Menu>
 
diff --git a/akregator/src/mainwindow.cpp b/akregator/src/mainwindow.cpp
index e23dd2a..fdc32e9 100644
--- a/akregator/src/mainwindow.cpp
+++ b/akregator/src/mainwindow.cpp
@@ -65,6 +65,10 @@ MainWindow::MainWindow( QWidget* parent, Qt::WindowFlags f )
     // set the shell's ui resource file
     setXMLFile("akregator_shell.rc");
 
+    KStandardAction::configureNotifications(m_part, SLOT(showKNotifyOptions()), actionCollection()); // options_configure_notifications
+    KStandardAction::keyBindings(this, SLOT(optionsConfigureKeys()), actionCollection()); // options_configure_keybinding
+    KStandardAction::configureToolbars(this, SLOT(optionsConfigureToolbars()), actionCollection()); // options_configure_toolbars
+
     toolBar()->show();
     statusBar()->show();
 
@@ -79,8 +83,6 @@ MainWindow::MainWindow( QWidget* parent, Qt::WindowFlags f )
     KStandardAction::showMenubar( menuBar(), SLOT(setVisible(bool)), actionCollection());
     setStandardToolBarMenuEnabled(true);
     createStandardStatusBarAction();
-
-    createGUI(0L);
 }
 
 bool MainWindow::loadPart()
