diff --git a/kdeui/xmlgui/kxmlguiclient.cpp b/kdeui/xmlgui/kxmlguiclient.cpp
index c5b391b..19f5028 100644
--- a/kdeui/xmlgui/kxmlguiclient.cpp
+++ b/kdeui/xmlgui/kxmlguiclient.cpp
@@ -229,7 +229,9 @@ void KXMLGUIClient::setXML( const QString &document, bool merge )
     QDomDocument doc;
     QString errorMsg;
     int errorLine, errorColumn;
-    bool result = doc.setContent(document, &errorMsg, &errorLine, &errorColumn);
+    // QDomDocument raises a parse error on empty document, but we accept no app-specific document,
+    // in which case you only get ui_standards.rc layout.
+    bool result = document.isEmpty() || doc.setContent(document, &errorMsg, &errorLine, &errorColumn);
     Q_ASSERT_X( result, "setXML", "The XML file that was set to KXMLGUIClient is invalid. Please check its correctness with xmllint" );
     if ( result ) { // kept for release mode
         setDOMDocument( doc, merge );
