diff --git a/khtml/rendering/render_layer.cpp b/khtml/rendering/render_layer.cpp
index 54115a6..bb99a1f 100644
--- a/khtml/rendering/render_layer.cpp
+++ b/khtml/rendering/render_layer.cpp
@@ -1139,12 +1139,8 @@ bool RenderLayer::nodeAtPoint(RenderObject::NodeInfo& info, int x, int y)
     int stx = m_x;
     int sty = m_y;
 
-#ifdef __GNUC__
-#warning HACK
-#endif
     if (renderer()->isCanvas()) {
-        stx += static_cast<RenderCanvas*>(renderer())->view()->contentsX();
-        sty += static_cast<RenderCanvas*>(renderer())->view()->contentsY();
+        static_cast<RenderCanvas*>(renderer())->view()->revertTransforms(stx, sty);
     }
 
     QRect damageRect(stx,sty, width(), height());
