diff --git a/apps/lib/konq/konq_operations.cpp b/apps/lib/konq/konq_operations.cpp
index ce4173e..855cace 100644
--- a/apps/lib/konq/konq_operations.cpp
+++ b/apps/lib/konq/konq_operations.cpp
@@ -185,6 +185,8 @@ void KonqOperations::_del( Operation method, const KUrl::List & _selectedUrls, C
         return;
     }
 
+    selectedUrls = KonqMimeData::simplifyUrlList( selectedUrls );
+
     if ( confirmation == SKIP_CONFIRMATION || askDeleteConfirmation( selectedUrls, method, confirmation, parentWidget() ) )
     {
         //m_srcUrls = selectedUrls;
diff --git a/apps/lib/konq/konqmimedata.cpp b/apps/lib/konq/konqmimedata.cpp
index 8a25586..f2c4355 100644
--- a/apps/lib/konq/konqmimedata.cpp
+++ b/apps/lib/konq/konqmimedata.cpp
@@ -22,14 +22,18 @@
 #include <kdebug.h>
 
 void KonqMimeData::populateMimeData( QMimeData* mimeData,
-                                     const KUrl::List& kdeURLs,
+                                     const KUrl::List& _kdeURLs,
                                      const KUrl::List& mostLocalURLs,
                                      bool cut )
 {
     mostLocalURLs.populateMimeData( mimeData );
 
+    KUrl::List kdeURLs( _kdeURLs );
     if ( !kdeURLs.isEmpty() )
     {
+        if ( cut ) {
+            kdeURLs = simplifyUrlList( kdeURLs );
+        }
         QMimeData tmpMimeData;
         kdeURLs.populateMimeData(&tmpMimeData);
         mimeData->setData("application/x-kde-urilist",tmpMimeData.data("text/uri-list"));
@@ -63,3 +67,20 @@ bool KonqMimeData::decodeIsCutSelection( const QMimeData *mimeData )
         return (a.at(0) == '1'); // true if 1
     }
 }
+
+KUrl::List KonqMimeData::simplifyUrlList( const KUrl::List &urls )
+{
+    KUrl::List ret( urls );
+    KUrl::List::const_iterator it = urls.begin();
+    while ( it != urls.end() ) {
+        KUrl::List::const_iterator it2 = urls.begin();
+        while ( it2 != urls.end() ) {
+            if ( it != it2 && it->isParentOf( *it2 ) ) {
+                ret.removeAll( *it2 );
+            }
+            ++it2;
+        }
+        ++it;
+    }
+    return ret;
+}
diff --git a/apps/lib/konq/konqmimedata.h b/apps/lib/konq/konqmimedata.h
index 9d4ede9..ea889c1 100644
--- a/apps/lib/konq/konqmimedata.h
+++ b/apps/lib/konq/konqmimedata.h
@@ -52,6 +52,18 @@ public:
      * @return true if the urls in @p mimeData were cut
      */
     static bool decodeIsCutSelection( const QMimeData *mimeData );
+
+    /**
+     * Remove urls from the list if an ancestor is present on the list. This can
+     * be used to delete only the ancestor url and skip a potential error of a non-existant url.
+     *
+     * For example, for a list of "/home/foo/a", "/home/foo/a/a.txt", "/home/foo/a/a/a.txt", "/home/foo/a/b/b.txt",
+     * "home/foo/b/b.txt", this method will return the list "/home/foo/a", "/home/foo/b/b.txt".
+     *
+     * @return the list @p urls without parented urls inside.
+     * @since 4.2
+     */
+    static KUrl::List simplifyUrlList( const KUrl::List & urls );
 };
 
 #endif /* KONQMIMEDATA_H */
