diff --git a/apps/konqueror/src/konqviewmanager.cpp b/apps/konqueror/src/konqviewmanager.cpp
index e9669fa..9c00b5a 100644
--- a/apps/konqueror/src/konqviewmanager.cpp
+++ b/apps/konqueror/src/konqviewmanager.cpp
@@ -62,13 +62,8 @@ KonqViewManager::KonqViewManager( KonqMainWindow *mainWindow )
   m_pamProfiles = 0L;
   m_bProfileListDirty = true;
   m_bLoadingProfile = false;
-
-  m_activePartChangedTimer = new QTimer(this);
-  m_activePartChangedTimer->setSingleShot(true);
-
   m_tabContainer = 0;
 
-  connect(m_activePartChangedTimer, SIGNAL(timeout()), this, SLOT( emitActivePartChanged()));
   connect( this, SIGNAL( activePartChanged ( KParts::Part * ) ),
            this, SLOT( slotActivePartChanged ( KParts::Part * ) ) );
 }
@@ -312,7 +307,8 @@ void KonqViewManager::breakOffTab( KonqFrameBase* currentFrame, const QSize& win
 
   KonqMainWindow *mainWindow = new KonqMainWindow(KUrl(), m_pMainWindow->xmlFile());
 
-  mainWindow->viewManager()->loadViewProfileFromGroup( profileGroup, QString() );
+  mainWindow->viewManager()->loadViewProfileFromGroup( profileGroup, currentProfile() );
+  mainWindow->applyMainWindowSettings( m_pMainWindow->autoSaveConfigGroup() );
 
   KonqFrameTabs * kft = mainWindow->viewManager()->tabContainer();
   KonqFrameBase *newFrame = dynamic_cast<KonqFrameBase*>(kft->currentWidget());
@@ -976,8 +972,6 @@ void KonqViewManager::setActivePart( KParts::Part *part, bool immediate )
     //if ( part )
     //    kDebug(1202) << part->metaObject()->className() << part->name();
 
-    // Due to the single-shot timer below, we need to also make sure that
-    // the mainwindow also has the right part active already
     KParts::Part* mainWindowActivePart = m_pMainWindow->currentView()
                                          ? m_pMainWindow->currentView()->part() : 0;
     if (part == activePart() && (!immediate || mainWindowActivePart == part))
@@ -1008,20 +1002,7 @@ void KonqViewManager::setActivePart( KParts::Part *part, bool immediate )
     if (part && part->widget())
         part->widget()->setFocus();
 
-    if (!immediate && reason() != ReasonRightClick) {
-        // We use a 0s single shot timer so that when left-clicking on a part,
-        // we process the mouse event before rebuilding the GUI.
-        // Otherwise, when e.g. dragging icons, the mouse pointer can already
-        // be very far from where it was...
-        // TODO: use a QTimer member var, so that if two conflicting calls to
-        // setActivePart(part,immediate=false) happen, the 1st one gets canceled.
-        m_activePartChangedTimer->start( 0 );
-        // This is not done with right-clicking so that the part is activated before the
-        // popup appears (#75201)
-    } else {
-        m_activePartChangedTimer->stop();
-        emitActivePartChanged();
-    }
+    emitActivePartChanged();
 }
 
 void KonqViewManager::slotActivePartChanged ( KParts::Part *newPart )
diff --git a/apps/konqueror/src/konqviewmanager.h b/apps/konqueror/src/konqviewmanager.h
index ba6965d..78977f5 100644
--- a/apps/konqueror/src/konqviewmanager.h
+++ b/apps/konqueror/src/konqviewmanager.h
@@ -414,8 +414,6 @@ private:
   QString m_currentProfileText;
 
     QMap<QString /*display name*/, QString /*path to file*/> m_mapProfileNames;
-
-  QTimer *m_activePartChangedTimer;
 };
 
 #endif
