diff --git a/okular/shell/shell.cpp b/okular/shell/shell.cpp
index 33f5fb7..4c9fb89 100644
--- a/okular/shell/shell.cpp
+++ b/okular/shell/shell.cpp
@@ -76,8 +76,8 @@ void Shell::init()
       // tell the KParts::MainWindow that this is indeed the main widget
       setCentralWidget(m_part->widget());
       // and integrate the part's GUI with the shell's
-      setupGUI(Keys | Save);
       createGUI(m_part);
+      setupGUI(Keys | Save);
       m_showToolBarAction = static_cast<KToggleAction*>(toolBarMenuAction());
       m_doc = qobject_cast<KDocumentViewer*>(m_part);
     }
