Index: kate/plugins/CMakeLists.txt
===================================================================
--- kate/plugins/CMakeLists.txt	(revisión: 758138)
+++ kate/plugins/CMakeLists.txt	(copia de trabajo)
@@ -7,4 +7,4 @@ add_subdirectory( wordcompletion )
 # This plugin is for howto-write-a-plugin purposes. It is not meant to be
 # installed. If you want to do so, just uncomment the add_subdirectory line.
 #
-# add_subdirectory( timedate )
+add_subdirectory( timedate )
Index: kate/dialogs/katedialogs.cpp
===================================================================
--- kate/dialogs/katedialogs.cpp	(revisión: 758138)
+++ kate/dialogs/katedialogs.cpp	(copia de trabajo)
@@ -825,12 +825,12 @@ void KatePartPluginConfigPage::apply ()
   KatePartPluginList &katePluginList = KatePartPluginManager::self()->pluginList();
   for (int i=0; i < plugins.count(); i++) {
     if (plugins[i].isPluginEnabled()) {
-      if (!(katePluginList[i].load && katePluginList[i].plugin)) {
+      if (!katePluginList[i].load) {
         KatePartPluginManager::self()->loadPlugin(katePluginList[i]);
         KatePartPluginManager::self()->enablePlugin(katePluginList[i]);
       }
     } else {
-      if (katePluginList[i].load || katePluginList[i].plugin) {
+      if (katePluginList[i].load) {
         KatePartPluginManager::self()->disablePlugin(katePluginList[i]);
         KatePartPluginManager::self()->unloadPlugin(katePluginList[i]);
       }
Index: kate/utils/katepartpluginmanager.cpp
===================================================================
--- kate/utils/katepartpluginmanager.cpp	(revisión: 758138)
+++ kate/utils/katepartpluginmanager.cpp	(copia de trabajo)
@@ -29,6 +29,7 @@
 #include <ktexteditor/view.h>
 #include <kconfig.h>
 #include <kconfiggroup.h>
+#include <kxmlguifactory.h>
 
 #include <kservicetypetrader.h>
 #include <kdebug.h>
@@ -184,6 +185,7 @@ void KatePartPluginManager::loadPlugin (
   if (item.plugin) return;
 
   item.plugin = KTextEditor::createPlugin (item.service, this);
+  Q_ASSERT(item.plugin);
   item.load = (item.plugin != 0);
 }
 
@@ -202,9 +204,22 @@ void KatePartPluginManager::enablePlugin
 
   // register docs and views
   foreach (KTextEditor::Document *doc, KateGlobal::self()->documents()) {
-    item.plugin->addDocument(doc);
-    foreach (KTextEditor::View *view, doc->views())
+    if (!doc)
+      continue;
+
+    foreach (KTextEditor::View *view, doc->views()) {
+      if (!view)
+        continue;
+
+      KXMLGUIFactory *viewFactory = view->factory();
+      if (viewFactory && view->isActiveView())
+        viewFactory->removeClient(view);
+
       item.plugin->addView(view);
+
+      if (viewFactory && view->isActiveView())
+        viewFactory->addClient(view);
+    }
   }
 }
 
@@ -216,9 +231,15 @@ void KatePartPluginManager::disablePlugi
 
   // de-register docs and views
   foreach (KTextEditor::Document *doc, KateGlobal::self()->documents()) {
-    foreach (KTextEditor::View *view, doc->views())
+    if (!doc)
+      continue;
+
+    foreach (KTextEditor::View *view, doc->views()) {
+      if (!view)
+        continue;
+
       item.plugin->removeView(view);
-    item.plugin->removeDocument(doc);
+    }
   }
 }
 
