Index: kutils/kpluginselector.cpp
===================================================================
--- kutils/kpluginselector.cpp	(revisión: 754930)
+++ kutils/kpluginselector.cpp	(copia de trabajo)
@@ -78,7 +78,6 @@
     pluginDelegate->setSeparatorPixels(8);
 
     QFont title(parent->font());
-    title.setPointSize(title.pointSize() + 2);
     title.setWeight(QFont::Bold);
 
     QFontMetrics titleMetrics(title);
@@ -806,7 +805,6 @@
     QFont previousFont(painter->font());
     QFont configureFont(painter->font());
 
-    title.setPointSize(title.pointSize() + 2);
     title.setWeight(QFont::Bold);
 
     if (index.internalPointer())
@@ -968,16 +966,15 @@
 
         QFont painterFont = painter->font();
         painterFont.setWeight(QFont::Bold);
-        painterFont.setPointSize(painterFont.pointSize() + 2);
         QFontMetrics metrics(painterFont);
         painter->setFont(painterFont);
 
         opt.fontMetrics = painter->fontMetrics();
 
         QRect auxRect(optionCopy.rect.left() + leftMargin,
-                      optionCopy.rect.bottom() - 2,
+                      optionCopy.rect.bottom() - 1,
                       optionCopy.rect.width() - leftMargin - rightMargin,
-                      2);
+                      1);
 
         QPainterPath path;
         path.addRect(auxRect);
@@ -1021,7 +1018,6 @@
     Q_UNUSED(option);
 
     QFont title(option.font);
-    title.setPointSize(title.pointSize() + 2);
     title.setWeight(QFont::Bold);
 
     QFontMetrics titleMetrics(title);
@@ -1042,7 +1038,7 @@
         return QSize(minimumItemWidth, qMax((separatorPixels * 2) + iconHeight, (separatorPixels * 3) + qMax(titleMetrics.height(), opt.rect.height()) + currentMetrics.height()));
     }
 
-    return QSize(minimumItemWidth, separatorPixels + titleMetrics.height() + 2);
+    return QSize(minimumItemWidth, separatorPixels + titleMetrics.height() + 1);
 }
 
 void KPluginSelector::Private::PluginDelegate::setIconSize(int width, int height)
